Enhancement/refactor randomisers

junniest opened the pull request:
Addressing #90 to remove traits/structs that have similar names to objects existing in rand already, removing unnecessary Mutex'es and simplifying the random number generators in general.

junniest committed:
Remove Mutex from rngs and make methods use &mut

Removed Mutexes from default and fake rngs and made methods use &mut self when necessary
instead

junniest committed:
Fix examples to use &mut rng

junniest committed:
Fix benches to use &mut rng

junniest committed:
Fix misleading comment about thread safety

junniest committed:
Refactor random number generation

Refactored random number generation:
* Removed unnecessary (and confusing) RandomSource trait
* Removed unnecessary Mutex from RandomGenerator

junniest committed:
Refactor FakeGenerator -> FakeRng

Refactor FakeGenerator -> FakeRng, now implementing the RngCore, SeedableRng and Send traits
that are needed to use it as a regular Rng. Additionally removed the Mutexes, unnecessary fields,
methods and tests.

junniest committed:
Remove cfg(test) annotation from FakeGenerator

Remove cfg(test) annotation from FakeGenerator since it is needed for
benchmarks.

junniest committed:
Update benches to use FakeGenerator

Update benchmark code to use FakeGenerator::default instead of new and FakeRng as
the random provider

junniest committed:
Update rng example

Updated rng example to output seeds from the rng itself and removed the
RandomSource trait use.

junniest committed:
Update phylo_info, tree and topo_opt with new rng

Updated phylo_info, tree and topo_optimiser with the new RandomGenerator and removed
the RandomSource trait use

junniest committed:
FakeGenerator::default instead of new in topo test

junniest committed:
Use FakeGenerator::default in tree tests

junniest committed:
Remove unnecessary rng use in tree node tests

junniest committed:
Added missing mut in test macro definition

junniest committed:
Upgrage rand crate 0.8.5 -> 0.9.0

junniest committed:
Rename gen(_range,_bool) to random, as in new rand

junniest committed:
Replace default StdRng with ChaCha8Rng

junniest committed:
Add comment on thread-wide rng irreproducibility

junniest commented:
Refactored randomisers, changed the default random to ChaCha8Rng for platform independence.

junniest committed:
Remove unnecessary type conversions

junniest committed:
Fix random doctest

junniest committed:
Update rand dep and remove unnecessary small_rng

junniest committed:
Remove unnecessary annotations from random

junniest committed:
fix rand version

junniest committed:
Add FakeRng test for coverage

junniest committed:
Add derives to RandomGenerator

junniest committed:
Add derives to FakeRng

junniest committed:
Add seed test for FakeRng

junniest committed:
Fix test comments

junniest committed:
Make FakeRng fake desired f64 values

junniest committed:
Replace ChaCha8Rng as default with faster Pcg32

junniest committed:
Fix FakeRng comments and test comments

junniest committed:
Add f64 value checks to reproducibility test

MattesMrzik commented on phylo/src/random/fake_random.rs:
```diff
@@ -1,477 +1,305 @@
+        let val: f64 = fake_rng.random();
+        assert_eq!(val, 0.0);
+        let val: bool = fake_rng.random();
+        assert!(!val);
```
Isnt this the same as `fake_rng_defaults` above?

junniest replied:
Yes, that's a glitch, thank you for catching it! I was trying to check both `FakeRng::default()` and `FakeGenerator::default()`, but identical tests are silly. Added `FakeGenerator::default();` to the previous test instead.

MattesMrzik commented on phylo/src/random/fake_random.rs:
```diff
@@ -1,477 +1,305 @@
         for i in 0..10 {
-            assert_eq!(fake_rng.gen_bool(0.7), values[i % values.len()]);
+            let val: u64 = fake_rng.random();
+            assert_eq!(val, values[i % values.len()]);
```
perhaps loop over 0..30, because otherwise the `i % values.len()` is never wrapped around.

MattesMrzik commented on phylo/src/random/fake_random.rs:
```diff
@@ -1,477 +1,305 @@
-        let val: u64 = fake_rng.gen();
-        assert_eq!(val, 0); // Default for u64
-        assert!(!fake_rng.gen::<bool>()); // Default for bool
+    fn fake_sample_default() {
+        // always returns first non-zero index when no pre-configured values
+        let mut rng = FakeGenerator::default();
+        let dist = WeightedIndex::new([1.0, 2.0, 3.0]).unwrap();
+        assert_eq!(rng.sample(&dist), 0);
```
Could we also make a similar test like 
```rust
let floats = vec![0.1, 0.25, 0.5, 0.75, 0.9];
let mut rng = RandomGenerator::from_rng(FakeRng::from_f64_values(floats));
let dist = WeightedIndex::new([1.0, 2.0, 3.0]).unwrap();
assert_eq!(rng.sample(&dist), ...);
```
and then assert to whatever value actually turns up?
just asking out of curiosity.

This might be useful to know in the tree builder. Since if we want to use this sampler to pick two nodes that should be joined next in the temperature version of the nj builder, it might be nicer to set up the fake rng to not always pick the first non-zero entry of the WeightedIndex.

junniest replied:
Yes, of course! I will add this test, that's my bad.

And you are actually correct, this was exactly the use case (`NjTreeBuilder`) I added this for, so that I can force the sampler to pick the entries I want. The previous version was a bit more straightforward because I could just force it to pick a specific entry in the vector by index, but this is not possible when you implement the Rng traits unfortunately.

MattesMrzik commented on phylo/src/random/mod.rs:
```diff
@@ -1,198 +1,145 @@
 /// // Create a custom RNG instance
-/// let custom_rng: RandomGenerator<StdRng> = RandomGenerator::new(123);
-/// let custom_value: f64 = custom_rng.gen();
+/// let mut custom_rng: RandomGenerator<StdRng> = RandomGenerator::new(123);
```
Here, you are using `StdRng,` but below the `Pcg32`, why this discrepancy?
```rust
pub type DefaultGenerator = RandomGenerator<Pcg32>;
```

junniest replied:
Ah, because I changed the main code and forgot to change the examples (; Ultimately it should not matter, the example just was to show that it is possible to use different rngs with the struct. But I will change the examples to use DefaultGenerator (which is Pcg32 now).

MattesMrzik commented on phylo/src/random/fake_random.rs:
```diff
@@ -1,477 +1,305 @@
-impl FakeGenerator {
-    /// Create a new FakeGenerator with empty value sequences,
-    /// which will default to 0, 0.0, or false if no values are provided
+unsafe impl Send for FakeRng {}
```
Is this because the compiler thinks sending FakeRng is unsafe, so we add the `unsafe` to tell the compiler that is in fact sendable?
But why would it be unsafe? its just a vector and a usize.
I also could delete this line without getting and error
I assume `Pcg32` also implements Send.

junniest replied:
This is curious, I think the requirement for `Send` (`R: Rng + SeedableRng + Send`) in `RandomGenerator` actually came from a Copilot suggestion that I did not check. I removed it now so the `impl` became unnecessary (at least at the moment).

However, the `Send` impl requires `unsafe` because of the trait definition, so the keyword would be necessary if the trait was required.

MattesMrzik commented on phylo/src/random/mod.rs:
```diff
@@ -1,198 +1,145 @@
 /// A generic random number generator wrapper that can work with different RNGs.
 ///
-/// This provides a thread-safe, reusable interface for different RNG backends.
+/// This provides a reusable interface for different RNG backends (not thread-safe per se).
```
Thread safe means that we don't necessarily get the same random values for different runs, right?
So if we have a rng with seed then up to creating the new thead it should be reproducible, right? If we then move or copy the rng to the thread, shouldn't the thread reproduce the same random numbers? Perhaps if I create multiple threads and some finish faster or slower between different runs then the order of how new threads are created is different between the runs, and therefore different rngs (or different states of a rng) are passed to the threads. Am I understanding this correctly?
And this has nothing to do with the `unsafe impl Send for FakeRng {}`, right?

junniest replied:
The problem with making the rng accessible from multiple threads at the same time is that the order in which the threads will poll it will make a difference to what the actual outcome of the analysis will be. I think that here my terminology is not right, it's not about thread safety per se, it's rather about computational reproducibility in the context of phylogenetics. I will rephrase that.

At the moment the threads that spawn for evaluating SPR moves are deterministic -- they actually do not need/use an rng. And I think that they are created in the same order every time, as far as I understand, but this might differ between different parallel implementations, I am not sure. If we ever need the threads to use an rng, an easy fix would be to give each thread a new rng instance seeded with a value from the "main" rng which would then maintain a reproducible sequence (but for this I would need to make sure the threads are always created in the same order). Although this will break the reproducibility between parallel/sequential runs with the same seed, so I would have to make similar modifications for the sequential implementation to make sure the SPR moves are evaluated exactly the same... But for now this is not a problem because the parallelisation doesn't need an rng in the first place.

And yes, this indeed has nothing to do with that `impl` which is now gone anyways (:

junniest replied:
Ah, again, this was an outdated comment. I added a note to the `DefaultGenerator` definition about this, but did not change the `RandomGenerator<R>` comment. I have now moved the note up, so now it says instead:
```
/// Note: This implementation is not thread-reproducible, i.e. if used in multiple threads, the sequences
/// generated may differ between runs. For phylogenetic analysis reproducibility both sequential and parallel
/// runs should produce the same results when using the same seed.
/// Users must handle this accordingly to ensure reproducibility if needed.
```

Does that make more sense to you?

MattesMrzik commented on phylo/examples/rng_examples.rs:
```diff
@@ -1,79 +1,89 @@
-    println!("1. RNG (StdRng) with seed {seed} - reproducible:");
-    let rng = DefaultGenerator::new(seed);
+    let mut rng = DefaultGenerator::new(seed);
+    println!("1. RNG (StdRng) with seed {} - reproducible:", rng.seed());
```
Aren't we using `Pcg32`?

junniest replied:
Yep, fixed now (:

MattesMrzik reviewed (approved):
Looks good to me 👍 
The `RandomGenerator` wrapper is just used to access the seed?
Some of the comments are just questions asking whether I am understanding sth correctly, we can also continue this discussion in person or via teams chat.

junniest committed:
Remove duplicate test

Removed duplicate test and added `FakeGenerator::default()` creatiob to the
`fake_rng_creation` test instead

junniest committed:
Increase loop over values in `FakeGenerator` test

Increased loop over values in `FakeGenerator` test so that it loops over predefined values
several times

junniest committed:
Clean up Rng examples

Clean up Rng examples to remove excessive duplication and use new DefaultGenerator which
is Pcg32-based instead of StdRng

junniest committed:
Remove `Send` requirement in RandomGenerator

Removed `Send` requirement in RandomGenerator as it is unnecessary (at least at the moment)

junniest committed:
Explain phylogenetic thread-wide reproducibility

Explained phylogenetic thread-wide reproducibility in a doctring associated with the
`RandomGenerator<R>`, previously was only written for the `DefaultGenerator` alias

junniest committed:
Add test for fake sampling an entry from a distr

Added test for fake sampling an entry from a distribution given the provided float values

junniest committed:
Add test_fill_bytes to FakeRng for coverage
